### DISCUSSION TITLE
keycloak-admin-client, Spring Boot 3, and Apache HttpClient dependency mismatch

### DISCUSSION THREAD

**Discussion by jopverhoeven** [keycloak/keycloak#24452] [Q&A]
Hi,

Recently I have been upgrading my application from Spring Boot 2.7.x to Spring Boot 3.0.x. With this upgrade the Apache HttpClient dependency in Spring Boot has been upgraded from [HttpClient4](https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient) to [HttpClient5](https://mvnrepository.com/artifact/org.apache.httpcomponents.client5/httpclient5).

When defining the Keycloak Admin Client I use the `resteasyClient` method to pass a HttpEngine with a custom HttpClient to the Keycloak Admin Client, but as the Keycloak Admin Client depends on [RESTEasy Client](https://mvnrepository.com/artifact/org.jboss.resteasy/resteasy-client/6.2.5.Final) (which in turn depends on the Apache HttpClient4) I can no longer pass the custom HttpClient to the Keycloak Admin Client.

```java
HttpClient httpClient5 = HttpClientBuilder.create().setConnectionManager(keycloakConnectionManager).build();
ApacheHttpClient43Engine engine = new ApacheHttpClient43Engine(httpClient5, false); <-- This line now goes wrong

keycloakAdminClient = KeycloakBuilder.builder()
                                     .realm("master")
                                     .serverUrl(keycloakProperties.getBaseUrl())
                                     .username(keycloakProperties.getAdminUsername())
                                     .password(keycloakProperties.getAdminPassword())
                                     .clientId("admin-cli")
                                     .resteasyClient(new ResteasyClientBuilderImpl().httpEngine(engine).build())
                                     .build();
```

Above I use the `ApacheHttpClient43Engine`, which requires a HttpClient from Apache HttpClient4. As there is no alternative to the `ApacheHttpCLient43Engine` for Apache HttpClient5 in the RESTEasy Client dependency of Keycloak Admin Client, I can no longer pass the custom HttpClient to Keycloak Admin Client.

The custom HttpClient uses Spring Boot specific code, which requires the Apache HttpClient5 as of Spring Boot 3, so downgrading is not an option sadly. 

Does anyone have suggestions on how to work around this, or do I simply have to wait until the Keycloak Admin Client depends on a version of RESTEasy Client which supports Apache HttpClient5?

**Comment by sRuklic-ag04** [comment]
did you manage to solve it?
